File: mobile_client/lib/pages/ondevice_ai_settings_page.dart
import 'dart:async';

import 'package:flutter/material.dart';

import '../services/local_ai_settings.dart';
import '../services/on_device_ai.dart';
import '../services/on_device_ai_types.dart';
import '../theme.dart';
import '../widgets/glass.dart';

/// Configuration for the on-device AI layer: tier, per-surface toggles, the
/// active local engine/model (+ downloads), tunables, and a debug generate box.
class OnDeviceAiSettingsPage extends StatefulWidget {
  const OnDeviceAiSettingsPage({super.key});

  @override
  State<OnDeviceAiSettingsPage> createState() => _OnDeviceAiSettingsPageState();
}

class _OnDeviceAiSettingsPageState extends State<OnDeviceAiSettingsPage> {
  final _s = LocalAiSettings.instance;
  final _ai = OnDeviceAi.instance;

  OnDeviceAvailability? _avail;
  List<LocalModelInfo> _models = const [];
  final Map<String, double> _downloading = {}; // id -> progress
  bool _loading = true;

  @override
  void initState() {
    super.initState();
    _refresh();
  }

  Future<void> _refresh() async {
    final avail = await _ai.availability();
    final models = await _ai.listModels();
    if (!mounted) return;
    setState(() {
      _avail = avail;
      _models = models;
      _loading = false;
    });
  }

  Future<void> _save() async {
    await _s.save();
    if (mounted) setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      extendBodyBehindAppBar: true,
      appBar: AppBar(
        title: const Text('On-device AI'),
        backgroundColor: Colors.transparent,
        centerTitle: true,
      ),
      body: AppBackground(
        child: SafeArea(
          child: ListView(
            padding: const EdgeInsets.fromLTRB(16, 8, 16, 32),
            children: [
              _availabilityCard(),
              glassSectionLabel('Mode'),
              GlassGroup(children: _tierRows()),
              glassSectionLabel('Use on-device for'),
              GlassGroup(children: [
                _switchTile('Chat', _s.chatEnabled, (v) {
                  setState(() => _s.chatEnabled = v);
                  _save();
                }),
                _switchTile('Voice', _s.voiceEnabled, (v) {
                  setState(() => _s.voiceEnabled = v);
                  _save();
                }),
              ]),
              glassSectionLabel('Local model'),
              GlassGroup(children: _modelRows()),
              glassSectionLabel('Advanced'),
              GlassGroup(children: _advancedRows()),
              glassSectionLabel('Test'),
              const _DebugGenerate(),
            ],
          ),
        ),
      ),
    );
  }

  Widget _availabilityCard() {
    final a = _avail;
    final ok = a?.available == true;
    return Padding(
      padding: const EdgeInsets.only(bottom: 8),
      child: GlassCard(
        child: Row(
          children: [
            Icon(ok ? Icons.check_circle_rounded : Icons.error_outline_rounded,
                color: ok ? JcTheme.cyan : JcTheme.muted),
            const SizedBox(width: 12),
            Expanded(
              child: Text(
                _loading
                    ? 'Checking on-device engine…'
                    : ok
                        ? 'On-device engine ready (${a?.engine ?? "?"})'
                        : 'Unavailable: ${a?.reason ?? "unknown"}',
                style: const TextStyle(color: JcTheme.text),
              ),
            ),
            IconButton(
              icon: const Icon(Icons.refresh_rounded, color: JcTheme.muted),
              onPressed: _refresh,
            ),
          ],
        ),
      ),
    );
  }

  List<Widget> _tierRows() {
    Widget tile(LocalAiTier t, String title, String sub) {
      final selected = _s.tier == t;
      return ListTile(
        leading: Icon(
          selected ? Icons.radio_button_checked : Icons.radio_button_unchecked,
          color: selected ? JcTheme.cyan : JcTheme.muted,
        ),
        title: Text(title, style: const TextStyle(color: JcTheme.text)),
        subtitle: Text(sub, style: const TextStyle(color: JcTheme.muted, fontSize: 12)),
        onTap: () {
          setState(() => _s.tier = t);
          _save();
        },
      );
    }
    return [
      tile(LocalAiTier.off, 'Off', 'Everything runs on the server.'),
      tile(LocalAiTier.routerCommands, 'Router + instant commands',
          'Answer trivial turns + fire device commands locally; escalate the rest.'),
      tile(LocalAiTier.fullLocalFirst, 'Full local-first',
          'Also attempt local tool calls; escalate only when needed.'),
    ];
  }

  List<Widget> _modelRows() {
    if (_models.isEmpty) {
      return [
        const ListTile(
          title: Text('No local models', style: TextStyle(color: JcTheme.text)),
          subtitle: Text('Apple Foundation Models needs iOS 26 + Apple Intelligence.',
              style: TextStyle(color: JcTheme.muted, fontSize: 12)),
        ),
      ];
    }
    return _models.map((m) {
      final selected = _s.activeLocalModelId == m.id;
      final dl = _downloading[m.id];
      return ListTile(
        leading: Icon(
          m.engine == 'apple-fm' ? Icons.apple : Icons.memory_rounded,
          color: selected ? JcTheme.cyan : JcTheme.muted,
        ),
        title: Text(m.label, style: const TextStyle(color: JcTheme.text)),
        subtitle: Text(
          dl != null
              ? 'Downloading ${(dl * 100).round()}%'
              : m.installed
                  ? (m.ramHintMb != null ? '~${m.ramHintMb} MB RAM' : 'Installed')
                  : 'Not downloaded${m.sizeBytes != null ? " · ${(m.sizeBytes! / 1e6).round()} MB" : ""}',
          style: const TextStyle(color: JcTheme.muted, fontSize: 12),
        ),
        trailing: _modelTrailing(m, selected, dl),
        onTap: m.installed
            ? () {
                setState(() => _s.activeLocalModelId = m.id);
                _save();
              }
            : null,
      );
    }).toList();
  }

  Widget _modelTrailing(LocalModelInfo m, bool selected, double? dl) {
    if (dl != null) {
      return const SizedBox(
          width: 22, height: 22, child: CircularProgressIndicator(strokeWidth: 2));
    }
    if (!m.installed) {
      return IconButton(
        icon: const Icon(Icons.download_rounded, color: JcTheme.cyan),
        onPressed: () => _download(m.id),
      );
    }
    return Row(mainAxisSize: MainAxisSize.min, children: [
      if (selected) const Icon(Icons.check_rounded, color: JcTheme.cyan),
      if (m.engine != 'apple-fm')
        IconButton(
          icon: const Icon(Icons.delete_outline_rounded, color: JcTheme.muted),
          onPressed: () async {
            await _ai.deleteModel(m.id);
            _refresh();
          },
        ),
    ]);
  }

  void _download(String id) {
    setState(() => _downloading[id] = 0);
    _ai.downloadModel(id).listen(
      (p) => setState(() => _downloading[id] = p),
      onError: (_) => setState(() => _downloading.remove(id)),
      onDone: () {
        setState(() => _downloading.remove(id));
        _refresh();
      },
    );
  }

  List<Widget> _advancedRows() {
    return [
      ListTile(
        title: const Text('Escalation deadline', style: TextStyle(color: JcTheme.text)),
        subtitle: Text('${_s.deadlineMs} ms — local attempt is dropped past this',
            style: const TextStyle(color: JcTheme.muted, fontSize: 12)),
      ),
      Slider(
        value: _s.deadlineMs.toDouble().clamp(200, 2000),
        min: 200,
        max: 2000,
        divisions: 18,
        activeColor: JcTheme.cyan,
        label: '${_s.deadlineMs} ms',
        onChanged: (v) => setState(() => _s.deadlineMs = v.round()),
        onChangeEnd: (_) => _save(),
      ),
      _switchTile('Confirm outward/destructive actions', _s.confirmLocalActions, (v) {
        setState(() => _s.confirmLocalActions = v);
        _save();
      }),
      _switchTile('Device commands short-circuit', _s.commandShortCircuit, (v) {
        setState(() => _s.commandShortCircuit = v);
        _save();
      }),
      _switchTile('Show "on-device" badge', _s.showBadge, (v) {
        setState(() => _s.showBadge = v);
        _save();
      }),
    ];
  }

  Widget _switchTile(String title, bool value, ValueChanged<bool> onChanged) {
    return SwitchListTile(
      value: value,
      onChanged: onChanged,
      activeThumbColor: JcTheme.cyan,
      title: Text(title, style: const TextStyle(color: JcTheme.text)),
    );
  }
}

/// A tiny on-device generate tester — types a prompt, streams the local model's
/// raw output (bypasses the router; pure engine check).
class _DebugGenerate extends StatefulWidget {
  const _DebugGenerate();

  @override
  State<_DebugGenerate> createState() => _DebugGenerateState();
}

class _DebugGenerateState extends State<_DebugGenerate> {
  final _ctrl = TextEditingController();
  String _out = '';
  StreamSubscription<String>? _sub;
  bool _running = false;

  @override
  void dispose() {
    _sub?.cancel();
    _ctrl.dispose();
    super.dispose();
  }

  void _run() {
    final text = _ctrl.text.trim();
    if (text.isEmpty || _running) return;
    setState(() {
      _out = '';
      _running = true;
    });
    final req = LocalRequest(
      userText: text,
      surface: VoiceSurface.chat,
      toolCatalogJson: '[]',
      tier: LocalAiTier.fullLocalFirst,
    );
    _sub = OnDeviceAi.instance.generate(req).listen(
      (t) => setState(() => _out += t),
      onError: (e) => setState(() {
        _out += '\n[error: $e]';
        _running = false;
      }),
      onDone: () => setState(() => _running = false),
    );
  }

  @override
  Widget build(BuildContext context) {
    return GlassCard(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          TextField(
            controller: _ctrl,
            style: const TextStyle(color: JcTheme.text),
            decoration: const InputDecoration(
              hintText: 'Prompt the local model…',
              hintStyle: TextStyle(color: JcTheme.muted),
              border: InputBorder.none,
            ),
            onSubmitted: (_) => _run(),
          ),
          const SizedBox(height: 8),
          Align(
            alignment: Alignment.centerRight,
            child: GlassButton(
              label: _running ? 'Running…' : 'Generate',
              onPressed: _running ? null : _run,
            ),
          ),
          if (_out.isNotEmpty) ...[
            const SizedBox(height: 12),
            Text(_out, style: const TextStyle(color: JcTheme.muted, fontSize: 13)),
          ],
        ],
      ),
    );
  }
}

File: ios_app/JarvisCopilot/Copilot/OnDeviceAI/OnDeviceAISettingsPage.swift
import SwiftUI

/// Configuration for the on-device AI layer: tier, per-surface toggles, the
/// active local model, tunables and a debug generate box.
///
/// Port of `mobile_client/lib/pages/ondevice_ai_settings_page.dart`. The model
/// Apple's model needs no download; an MLX row downloads on tap (with a
/// cancellable progress ring) and offers "Delete weights" on long-press.
struct OnDeviceAISettingsPage: View {
    @State private var store: OnDeviceAISettingsStore

    /// See `SettingsPage.init` — a view's `init` isn't main-actor-isolated, so
    /// the store can't be a default argument.
    init(store: OnDeviceAISettingsStore? = nil) {
        _store = State(initialValue: store
                       ?? MainActor.assumeIsolated { OnDeviceAISettingsStore() })
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 26) {
                availabilityCard
                section("Mode") { tierRows }
                section("Use on-device for") { surfaceRows }
                section("Local model") { modelRows }
                section("Advanced") { advancedRows }
                section("Test") { OnDeviceDebugGenerate(store: store) }
            }
            .padding(.horizontal, 20)
            .padding(.top, 8)
            .padding(.bottom, 32)
        }
        .scrollDismissesKeyboard(.interactively)
        .jcScreen("On-device AI")
        .task { await store.load() }
    }

    private func section<Content: View>(_ title: String,
                                        @ViewBuilder content: () -> Content) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            GlassSectionLabel(title)
            content()
        }
    }

    // MARK: - Availability

    private var availabilityCard: some View {
        GlassCard {
            HStack(spacing: 12) {
                Image(systemName: store.isReady ? "checkmark.circle.fill" : "exclamationmark.circle")
                    .font(.system(size: 22))
                    .foregroundStyle(store.isReady ? JcTheme.cyan : JcTheme.muted)
                Text(store.availabilitySummary)
                    .font(JcText.body)
                    .foregroundStyle(JcTheme.text)
                    .fixedSize(horizontal: false, vertical: true)
                Spacer(minLength: 8)
                Button {
                    Task { await store.refresh() }
                } label: {
                    Image(systemName: "arrow.clockwise")
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundStyle(JcTheme.muted)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Refresh")
            }
        }
    }

    // MARK: - Tier

    private var tierRows: some View {
        GlassGroup {
            tierRow(.off, "Off", "Everything runs on the server.")
            tierRow(.routerCommands, "Router + instant commands",
                    "Answer trivial turns + fire device commands locally; escalate the rest.")
            tierRow(.fullLocalFirst, "Full local-first",
                    "Also attempt local tool calls; escalate only when needed.", last: true)
        }
    }

    private func tierRow(_ tier: LocalAiTier, _ title: String, _ subtitle: String,
                         last: Bool = false) -> some View {
        let selected = store.settings.tier == tier
        return GlassRow(symbol: selected ? "largecircle.fill.circle" : "circle",
                        title: title, subtitle: subtitle, subtitleLineLimit: 3,
                        last: last, action: { store.setTier(tier) }) {
            if selected {
                Image(systemName: "checkmark")
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(JcTheme.cyan)
            }
        }
    }

    // MARK: - Surfaces

    private var surfaceRows: some View {
        GlassGroup {
            OnDeviceSwitchRow(symbol: "bubble.left", title: "Chat",
                              isOn: store.settings.chatEnabled) { store.setChatEnabled($0) }
            OnDeviceSwitchRow(symbol: "waveform", title: "Voice",
                              isOn: store.settings.voiceEnabled,
                              last: true) { store.setVoiceEnabled($0) }
        }
    }

    // MARK: - Models

    private var modelRows: some View {
        GlassGroup {
            if store.models.isEmpty {
                GlassRow(symbol: "cpu", title: "No local models",
                         subtitle: "Apple Foundation Models needs iOS 26 + Apple Intelligence.",
                         subtitleLineLimit: 3, last: true) { EmptyView() }
            } else {
                ForEach(Array(store.models.enumerated()), id: \.element.id) { index, model in
                    let selected = store.activeModelID == model.id
                    let progress = store.downloadProgress[model.id]
                    let downloadable = model.engine == .mlx && !model.installed
                    GlassRow(symbol: model.engine.symbol,
                             title: model.label,
                             subtitle: progress.map { "Downloading… \(Int($0 * 100))%" } ?? model.detail,
                             subtitleLineLimit: 2,
                             last: index == store.models.count - 1,
                             action: model.installed ? { store.selectModel(model) }
                                   : (downloadable && progress == nil) ? { store.download(model) } : nil) {
                        if let progress {
                            Button { store.cancelDownload(model) } label: {
                                ZStack {
                                    ProgressView(value: progress).progressViewStyle(.circular).tint(JcTheme.cyan)
                                    Image(systemName: "xmark").font(.system(size: 8, weight: .bold))
                                        .foregroundStyle(JcTheme.muted)
                                }
                                .frame(width: 22, height: 22)
                            }
                            .buttonStyle(.plain)
                        } else if selected {
                            Image(systemName: "checkmark")
                                .font(.system(size: 14, weight: .semibold))
                                .foregroundStyle(JcTheme.cyan)
                        } else if downloadable {
                            Image(systemName: "arrow.down.circle")
                                .font(.system(size: 16, weight: .medium))
                                .foregroundStyle(JcTheme.cyan)
                        }
                    }
                    .opacity(model.installed || downloadable ? 1 : 0.5)
                    .contextMenu {
                        if model.engine == .mlx, model.installed {
                            Button(role: .destructive) { store.delete(model) } label: {
                                Label("Delete weights", systemImage: "trash")
                            }
                        }
                    }
                }
                if let error = store.downloadError {
                    Text(error).font(.caption).foregroundStyle(JcTheme.danger).padding(.horizontal, 14).padding(.bottom, 8)
                }
            }
        }
    }

    // MARK: - Advanced

    private var advancedRows: some View {
        GlassGroup {
            GlassRow(symbol: "arrow.triangle.branch",
                     title: "The model decides escalation",
                     subtitle: "Each turn runs the on-device model to completion; it chooses "
                             + "to answer, act, or hand off to the server. No time limit.",
                     subtitleLineLimit: 4) { EmptyView() }
            OnDeviceSwitchRow(symbol: "exclamationmark.shield",
                              title: "Confirm outward/destructive actions",
                              isOn: store.settings.confirmLocalActions) {
                store.setConfirmLocalActions($0)
            }
            OnDeviceSwitchRow(symbol: "bolt",
                              title: "Device commands short-circuit",
                              isOn: store.settings.commandShortCircuit) {
                store.setCommandShortCircuit($0)
            }
            OnDeviceSwitchRow(symbol: "seal",
                              title: "Show \"on-device\" badge",
                              isOn: store.settings.showBadge,
                              last: true) { store.setShowBadge($0) }
        }
    }
}

/// A toggle row styled like `GlassRow` — the on-device screen's own copy, because
/// `SettingsPage`'s is file-private there.
private struct OnDeviceSwitchRow: View {
    let symbol: String
    let title: String
    var subtitle: String? = nil
    let isOn: Bool
    var last: Bool = false
    let onChange: (Bool) -> Void

    var body: some View {
        GlassRow(symbol: symbol, title: title, subtitle: subtitle,
                 subtitleLineLimit: 3, last: last) {
            Toggle("", isOn: Binding(get: { isOn }, set: onChange))
                .labelsHidden()
                .tint(JcTheme.primaryBlue)
        }
    }
}

/// A tiny on-device generate tester: type a prompt, stream the local model's raw
/// output. Bypasses the router — a pure engine check.
private struct OnDeviceDebugGenerate: View {
    @Bindable var store: OnDeviceAISettingsStore

    var body: some View {
        GlassCard {
            VStack(alignment: .leading, spacing: 12) {
                TextField("Prompt the local model…", text: $store.prompt, axis: .vertical)
                    .font(JcText.body)
                    .foregroundStyle(JcTheme.text)
                    .textFieldStyle(.plain)
                    .lineLimit(1...4)
                    .onSubmit { store.runDebugGenerate() }

                HStack {
                    Spacer(minLength: 0)
                    if store.running {
                        GlassButton(title: "Stop", ghost: true) { store.cancelDebugGenerate() }
                    }
                    GlassButton(title: store.running ? "Running…" : "Generate") {
                        store.runDebugGenerate()
                    }
                }

                if !store.output.isEmpty {
                    Text(store.output)
                        .font(JcText.small)
                        .foregroundStyle(JcTheme.muted)
                        .textSelection(.enabled)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
            }
        }
    }
}
